src/image-viewer.js

```diff
@@ -0,0 +1,648 @@
+      if (this.ariaAttributes_[key]) {
+        this.image_.setAttribute(key, this.ariaAttributes_[key]);
+      }
+    });
+
+    if (sourceImage && isLoaded(sourceImage) && sourceImage.src) {
+      // Set src provisionally to the known loaded value for fast display.
+      // It will be updated later.
+      this.image_.setAttribute('src', sourceImage.src);
+    }
+  }
+
+  /**
+   * Measures the image viewer and image sizes and positioning.
+   * @return {!Promise}
+   */
+  measure() {
+    this.viewerBox_ = layoutRectFromDomRect(this.viewer_
+        ./*OK*/getBoundingClientRect());
+
+    const sf = Math.min(this.viewerBox_.width / this.sourceWidth_,
+        this.viewerBox_.height / this.sourceHeight_);
+    let width = Math.min(this.sourceWidth_ * sf, this.viewerBox_.width);
+    let height = Math.min(this.sourceHeight_ * sf, this.viewerBox_.height);
+
+    // TODO(dvoytenko): This is to reduce very small expansions that often
+    // look like a stutter. To be evaluated if this is still the right
+    // idea.
+    if (width - this.sourceWidth_ <= 16) {
+      width = this.sourceWidth_;
+      height = this.sourceHeight_;
+    }
+
+    this.imageBox_ = layoutRectLtwh(
+        Math.round((this.viewerBox_.width - width) / 2),
+        Math.round((this.viewerBox_.height - height) / 2),
+        Math.round(width),
+        Math.round(height));
+
+    st.setStyles(this.image_, {
+      top: st.px(this.imageBox_.top),
+      left: st.px(this.imageBox_.left),
+      width: st.px(this.imageBox_.width),
+      height: st.px(this.imageBox_.height),
+    });
+
+    // Reset zoom and pan.
+    this.startScale_ = this.scale_ = 1;
+    this.startX_ = this.posX_ = 0;
+    this.startY_ = this.posY_ = 0;
+    this.updatePanZoomBounds_(this.scale_);
+    this.updatePanZoom_();
+
+    return this.updateSrc_();
+  }
+
+  /**
+   * @return {!Promise}
+   * @private
+   */
+  updateSrc_() {
+    if (!this.srcset_) {
+      // Do not update source if the lightbox has already exited.
+      return Promise.resolve();
+    }
+    this.maxSeenScale_ = Math.max(this.maxSeenScale_, this.scale_);
+    const width = this.imageBox_.width * this.maxSeenScale_;
+    const src = this.srcset_.select(width, this.lightbox_.getDpr()).url;
+    if (src == this.image_.getAttribute('src')) {
+      return Promise.resolve();
+    }
+    // Notice that we will wait until the next event cycle to set the "src".
+    // This ensures that the already available image will show immediately
+    // and then naturally upgrade to a higher quality image.
+    return Services.timerFor(this.win).promise(1).then(() => {
+      this.image_.setAttribute('src', src);
+      return this.loadPromise_(this.image_);
+    });
+  }
+
+  /** @private */
+  setupGestures_() {
+    const gestures = Gestures.get(this.image_, true);
+
+    // Movable.
+    gestures.onGesture(SwipeXYRecognizer, e => {
+      if (this.isScaled()) {
+        event.preventDefault();
```

Comment:
This logic was added. We only want to act on the SwipeXY if our image is scaled. If it is scaled, we also want to preventDefault to avoid the`amp-carousel` changing slides.